### PR TITLE
Feature/submit for review

### PR DESCRIPTION
**Pull request by romanetar**
ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3310526&stafilter=NotComplete&fileview=grid

ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3310529&stafilter=NotComplete&fileview=grid

ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3310527&stafilter=NotComplete&fileview=grid

ref https://tipit.avaza.com/project/view#!tab=task-pane&groupby=MyTaskDueDate&view=vertical&task=3310460&stafilter=NotComplete&fileview=grid

ref https://tipit.avaza.com/task#!sortby=DateUpdatedDesc&task=3310534&stafilter=NotComplete

Signed-off-by: romanetar <roman_ag@hotmail.com>

### CODE REVIEW

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEmailEventFlowApiController.php [hunk 1]
```diff
@@ -80,6 +80,7 @@ function getUpdateValidationRules(array $payload): array
     {
         return [
             'email_template_identifier' => 'sometimes|required|string',
+            'recipient'                 => 'sometimes|string',
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEmailEventFlowApiController.php:83]: add email validation and 255 max

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php [hunk 2]
```diff
@@ -216,4 +226,97 @@ public function publishAll($summit_id, $source)
...
+     */
+    public function addReview($summit_id, $source, $track_id)
+    {
+        return $this->processRequest(function () use ($summit_id, $source, $track_id) {
+            $summit = SummitFinderStrategyFactory::build($this->summit_repository, $this->resource_server_context)->find(intval($summit_id));
+            if (is_null($summit)) return $this->error404();
+
+            $payload = $this->getJsonPayload(ProposedScheduleLockValidationRulesFactory::buildForAdd());
+
+            $schedule = $this->service->addReview($summit, $source, intval($track_id), $payload);
```
Comment (smarcet) [inline app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php:245]: same as above

File: app/Jobs/Emails/ProposedSchedule/UnsubmitForReviewEmail.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        $submitter = $lock->getCreatedBy()->getMember();
+        $payload = [];
+
+        $payload['summit_name']         = $summit->getName();
+        $payload['summit_logo']         = $summit->getLogoUrl();
+        $payload['submitter_fullname']  = $submitter->getFullName();
+        $payload['submitter_email']     = $submitter->getEmail();
+        $payload['track']               = $lock->getTrack()->getTitle();
+        $payload['track_id']            = $lock->getTrack()->getId();
+        $payload['message']             = $lock->getReason();
```
Comment (smarcet) [inline app/Jobs/Emails/ProposedSchedule/UnsubmitForReviewEmail.php:39]: the message is not the one set originally
is the response from admin so it should be sent on the constructor as a param

File: app/Services/Model/Imp/ScheduleService.php [hunk 4]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
     }
+
+    /**
+     * @inheritDoc
+     */
+    public function addReview(Summit $summit, string $source, int $track_id, array $payload): SummitProposedSchedule
+    {
+        return $this->tx_service->transaction(function () use ($summit, $source, $track_id, $payload) {
+
+            $former_lock = $this->schedule_lock_repository->getBySummitAndTrackId($summit->getId(), $track_id);
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:435]: please check that track id belongs to summit first

File: app/Services/Model/Imp/ScheduleService.php [hunk 5]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+
+            if (!is_null($former_lock)) {
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:437]: is there is a former lock , then throw validation exception , not email, email is only for successful creation

File: app/Services/Model/Imp/ScheduleService.php [hunk 6]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+            }
+
+            $schedule = $this->schedule_repository->getBySourceAndSummitId($source, $summit->getId());
+
+            if (!$schedule instanceof SummitProposedSchedule) {
+                throw new EntityNotFoundException("can not find a proposed schedule for summit id {$summit->getId()} and source {$source}");
+            }
+            $track = $summit->getPresentationCategory($track_id);
+
+            if (!$track instanceof PresentationCategory) {
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:449]: move this validation up

File: app/Services/Model/Imp/ScheduleService.php [hunk 7]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+                throw new EntityNotFoundException("can not find a presentation category with id {$track_id} for summit id {$summit->getId()}");
+            }
+            $member = ResourceServerContext::getCurrentUser(false);
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:452]: dont get current member here, pass it by param and get it on controller

File: app/Services/Model/Imp/ScheduleService.php [hunk 8]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+            $lock = new SummitProposedScheduleLock();
+
+            if (isset($payload['message']))
+                $lock->setReason($payload['message']);
+
+            $lock->setTrack($track);
+            $lock->setCreatedBy($track_chair);
+
+            $schedule->addProposedScheduleLock($lock);
+            return $schedule;
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:467]: email should be sent here

File: app/Services/Model/Imp/ScheduleService.php [hunk 9]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function removeReview(Summit $summit, string $source, int $track_id): SummitProposedSchedule
+    {
+        return $this->tx_service->transaction(function () use ($summit, $source, $track_id) {
+
+            $lock = $this->schedule_lock_repository->getBySummitAndTrackId($summit->getId(), $track_id);
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:478]: check track belongs to summit

File: app/Services/Model/Imp/ScheduleService.php [hunk 10]
```diff
@@ -403,4 +424,68 @@ public function publishAll(string $source, int $summit_id, array $payload, ?Filt
...
+
+            $schedule->addProposedScheduleLock($lock);
+            return $schedule;
+        });
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function removeReview(Summit $summit, string $source, int $track_id): SummitProposedSchedule
```
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:474]: u need here the optional message from admin as a param

File: database/migrations/model/Version20230616161141.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+                $table->string('Reason', 1024)->setNotnull(false)->setDefault('');
+
+                // FK
+
+                $table->integer("SummitProposedScheduleID", false, true)->setNotnull(true);
+                $table->index("SummitProposedScheduleID", "SummitProposedScheduleID");
+                $table->foreign("SummitProposedSchedule", "SummitProposedScheduleID", "ID", ["onDelete" => "CASCADE"], "FK_SummitProposedScheduleLock_SummitProposedSchedule");
+
+                $table->integer("TrackID", false, false)->setNotnull(true);
```
Comment (smarcet) [inline database/migrations/model/Version20230616161141.php:52]: u need to add an unique filter by SummitProposedScheduleID and TrackID

**Review by smarcet — CHANGES_REQUESTED** [review]
@romanetar please review comments

File: app/Jobs/Emails/ProposedSchedule/SubmitForReviewEmail.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        $submitter = $lock->getCreatedBy()->getMember();
+        $payload = [];
+
+        $payload['summit_name']         = $summit->getName();
+        $payload['summit_logo']         = $summit->getLogoUrl();
+        $payload['submitter_fullname']  = $submitter->getFullName();
+        $payload['submitter_email']     = $submitter->getEmail();
+        $payload['track']               = $lock->getTrack()->getTitle();
+        $payload['track_id']            = $lock->getTrack()->getId();
+        $payload['message']             = $lock->getReason();
```
Comment (smarcet) [inline app/Jobs/Emails/ProposedSchedule/SubmitForReviewEmail.php:39]: add ?? ""

File: app/Services/Model/Imp/ScheduleService.php (hunk 4 above)
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:435]: do we need to check on repository to obtain the former lock ? can we just use the schedule to search for a lock by track id ?

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEmailEventFlowApiController.php (hunk 1 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitEmailEventFlowApiController.php:83]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 9 above)
Comment (smarcet) [inline app/Services/Model/Imp/ScheduleService.php:478]: get the lock from schedule , not from repo

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php [hunk 13]
```diff
@@ -216,4 +226,97 @@ public function publishAll($summit_id, $source)
...
     }
+
+
+    /**
+     * @param $summit_id
+     * @param $source
+     * @param $track_id
+     * @return \Illuminate\Http\JsonResponse|mixed
+     */
+    public function addReview($summit_id, $source, $track_id)
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php:237]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php (hunk 2 above)
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php:245]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php [hunk 14]
```diff
@@ -216,4 +226,97 @@ public function publishAll($summit_id, $source)
...
+    {
+        return $this->processRequest(function () use ($summit_id, $source, $track_id) {
+            $summit = SummitFinderStrategyFactory::build($this->summit_repository, $this->resource_server_context)->find(intval($summit_id));
+            if (is_null($summit)) return $this->error404();
+
+            $payload = $this->getJsonPayload(ProposedScheduleLockValidationRulesFactory::buildForAdd());
+
+            $schedule = $this->service->addReview($summit, $source, intval($track_id), $payload);
+
+            return $this->updated(SerializerRegistry::getInstance()->getSerializer($schedule)
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php:247]: done

File: app/Models/Foundation/Summit/EmailFlows/SummitEmailEventFlow.php [hunk 15]
```diff
@@ -83,4 +89,16 @@ public function setEventType(SummitEmailEventFlowType $event_type): void
         $this->event_type = $event_type;
     }
 
+    public function getEmailRecipient(): string
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Models/Foundation/Summit/EmailFlows/SummitEmailEventFlow.php:92]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:435]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 6 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:449]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 10 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:474]: done

File: app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php [hunk 16]
```diff
@@ -216,4 +226,97 @@ public function publishAll($summit_id, $source)
...
+     * @param $summit_id
+     * @param $source
+     * @param $track_id
+     * @return \Illuminate\Http\JsonResponse|mixed
+     */
+    public function removeReview($summit_id, $source, $track_id)
+    {
+        return $this->processRequest(function () use ($summit_id, $source, $track_id) {
+
+            $summit = SummitFinderStrategyFactory::build($this->summit_repository, $this->resource_server_context)->find(intval($summit_id));
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Http/Controllers/Apis/Protected/Summit/OAuth2SummitProposedScheduleApiController.php:267]: done

File: app/Jobs/Emails/ProposedSchedule/SubmitForReviewEmail.php (hunk 12 above)
Comment (romanetar) [reply app/Jobs/Emails/ProposedSchedule/SubmitForReviewEmail.php:39]: done

File: database/migrations/model/Version20230616161141.php (hunk 11 above)
Comment (romanetar) [reply database/migrations/model/Version20230616161141.php:52]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 9 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:478]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 4 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:435]: done

File: app/Models/Foundation/Summit/Repositories/ISummitProposedScheduleLockRepository.php [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+        Order $order   = null
+    );
+
+    /**
+     * @param int $summit_id
+     * @param int $track_id
+     * @return SummitProposedScheduleLock|null
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     */
+    public function getBySummitAndTrackId(int $summit_id, int $track_id): ?SummitProposedScheduleLock;
```
[the remark this answers is not in the document]
Comment (romanetar) [reply app/Models/Foundation/Summit/Repositories/ISummitProposedScheduleLockRepository.php:50]: removed

File: app/Services/Model/Imp/ScheduleService.php (hunk 8 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:467]: done

File: app/Services/Model/Imp/ScheduleService.php (hunk 5 above)
Comment (romanetar) [reply app/Services/Model/Imp/ScheduleService.php:437]: done

File: app/Jobs/Emails/ProposedSchedule/UnsubmitForReviewEmail.php (hunk 3 above)
Comment (romanetar) [reply app/Jobs/Emails/ProposedSchedule/UnsubmitForReviewEmail.php:39]: done

**Review by smarcet — COMMENTED** [review]
LGTM

**Pull request merged**